webapi/src/test/resources/test-data/ontologyR2RV2/bildSimple.json

```diff
@@ -12,79 +12,93 @@
         "rdfs:subClassOf" : [ "http://api.knora.org/ontology/knora-api/simple/v2#StillImageRepresentation", {
           "@type" : "owl:Restriction",
           "owl:cardinality" : 1,
-          "owl:onProperty" : "http://0.0.0.0:3333/ontology/00FF/images/simple/v2#bearbeiter"
+          "owl:onProperty" : "http://api.knora.org/ontology/knora-api/simple/v2#creationDate"
```

Comment:
Just a general question, where is this URL defined? I didn't find it in `application.conf`.  I don't think that this a good idea, as this would mean, that we would need to have something running or serving at this address? Why not have it like this: `http://0.0.0.0:3333/ontology/knora-api/simple/v2#creationDate"?

Reply:
This is documented here:

http://www.knora.org/documentation/manual/rst/03-knora-api-server/api_v2/knora-iris.html#external-ontology-iris

The idea is that we will have a Knora instance serving the built-in ontologies at that address.

If it was `http://0.0.0.0:3333/ontology/knora-api/simple/v2#creationDate`, then the `knora-api` ontology running on project-specific server A would not appear to be the same as the `knora-api` ontology running on project-specific server B. From a client's point of view, they would not be the same ontology so would not be interoperable.

Reply:
Would it be generally possible to "dump" the `knora-api` ontologies to files? I'm thinking of an easy way to serve all our ontologies. If all the ontologies would be available as files, then it wouldn't be difficult to serve them.

Reply:
The `knora-api` ontologies are already served by the Knora API server. You can request them from the API, download them, save them in files. That's what I did to make the test data for `OntologyV2R2RSpec`. It uses these files:

* The `knora-api` v2 ontology in the simple schema: https://github.com/dhlab-basel/Knora/blob/develop/webapi/src/test/resources/test-data/ontologyR2RV2/knoraApiOntologySimple.json
* The `knora-api` v2 ontology in the default schema: https://github.com/dhlab-basel/Knora/blob/develop/webapi/src/test/resources/test-data/ontologyR2RV2/knoraApiWithValueObjects.json

Reply:
Currently the API server only serves ontologies in JSON-LD. What I've been planning to do is get the API server to take an HTTP `Accept` header and then be able to convert the JSON-LD to Turtle, XML, or HTML.

Reply:
ok, great.

The reasoning behind me asking, is that (as far as I understand) all our ontologies aren't allowed to change (once they are published). So no reason to require a full running Knora stack, just to serve basically static content. Makes the life on the operations side much easier. Just keep in mind, that we don't (and probably never will) have a dedicated operations team. The less dependency there is on a running Knora stack, the less will break when the server is down.

Maybe then changing it to something like `http://www.knora.org/ontology/knora-api/simple/v2#creationDate` would allow them to be published together with the other Knora ontologies.

Reply:
We could change it to `www.knora.org`. Or we could just have something like `nginx` serve flat files at `api.knora.org/ontology`.

Reply:
I had the idea to put them on the `www.knora.org` site served from Github. With Nginx (or whatever), we would need to put them on our servers. We will see. I have opened an issue ##753 for this. I need to think this a bit trough. Anyway, we have some time until V2 is officially released.

Thanks for bearing with me :-)

Reply:
With the current design, we can't use `www.knora.org` for external (API v2) ontologies, because it's reserved for internal ontologies (the ones used in the triplestore):

http://www.knora.org/documentation/manual/rst/03-knora-api-server/api_v2/knora-iris.html#internal-ontology-iris

`SmartIriImpl` uses the hostname to determine whether the IRI is internal or external.

We could redesign all this, but I'd really rather not. :)

> Thanks for bearing with me :-)

No problem. :)

Reply:
Also, I don't think we should rely on GitHub to serve Knora ontologies. That would mean that if GitHub is down, the DaSCH is down. Also, my understanding is that as critical research infrastructure for Switzerland, the DaSCH is supposed to be hosted in Switzerland.

Reply:
> The less dependency there is on a running Knora stack

The DaSCH can't exist without a running Knora stack.

Reply:
> We could redesign all this, but I'd really rather not. :)

ok, we should then leave it as it is.

>> The less dependency there is on a running Knora stack

> The DaSCH can't exist without a running Knora stack.

What I meant by this is, if our Knora stack is down for a few hours (which is acceptable by DaSCH's / Lukas's standard) then this shouldn't have a negative impact on all other running Knora instances around Switzerland / World because they all depend on our served ontologies.

That's why I would like to see a separation of the serving of ontologies from the Knora stack and implement the serving of instances in a lightweight and highly available (redundant) way. Nginx serving flat files is a much simpler and lightweight (read costs less) setup, then the whole Knora stack.

We will host the DaSCH Knora stack with Switch, but not completely in a high availability way. The data will have a redundancy of 6 with data in Zürich and Lausanne (losing data is not acceptable), but the machine running the stack will not be redundant (few hours of downtime acceptable).

I would love to have everything highly redundant so that we don't have a minute of downtime, but the budget constraints are what they are. Well, maybe someday this will change.

Reply:
OK! I like flat files.